File: dotnet/src/Microsoft.Agents.Protocol.Hosting/Storage/InMemoryStorage.cs
using System.Collections.Concurrent;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Agents.Protocol.Hosting.Core;

namespace Microsoft.Agents.Protocol.Hosting.Storage;

/// <summary>
/// In-memory implementation of IStorage for development and testing.
/// </summary>
public class InMemoryStorage : IStorage
{
    private readonly ConcurrentDictionary<string, ConcurrentDictionary<string, object>> _storage = new();

    /// <inheritdoc/>
    public Task<T?> GetAsync<T>(string threadId, string key, CancellationToken cancellationToken = default)
    {
        if (cancellationToken.IsCancellationRequested)
        {
            return Task.FromCanceled<T?>(cancellationToken);
        }

        if (_storage.TryGetValue(threadId, out var threadData) &&
            threadData.TryGetValue(key, out var value))
        {
            return Task.FromResult<T?>((T)value);
        }

        return Task.FromResult<T?>(default);
    }

    /// <inheritdoc/>
    public Task SetAsync<T>(string threadId, string key, T value, CancellationToken cancellationToken = default)
    {
        if (cancellationToken.IsCancellationRequested)
        {
            return Task.FromCanceled(cancellationToken);
        }

        var threadData = _storage.GetOrAdd(threadId, _ => new ConcurrentDictionary<string, object>());
        threadData[key] = value!;
        return Task.CompletedTask;
    }

    /// <inheritdoc/>
    public Task DeleteAsync(string threadId, string key, CancellationToken cancellationToken = default)
    {
        if (cancellationToken.IsCancellationRequested)
        {
            return Task.FromCanceled(cancellationToken);
        }

        if (_storage.TryGetValue(threadId, out var threadData))
        {
            threadData.TryRemove(key, out _);
        }

        return Task.CompletedTask;
    }

    /// <inheritdoc/>
    public Task<string[]> GetKeysAsync(string threadId, CancellationToken cancellationToken = default)
    {
        if (cancellationToken.IsCancellationRequested)
        {
            return Task.FromCanceled<string[]>(cancellationToken);
        }

        if (_storage.TryGetValue(threadId, out var threadData))
        {
            return Task.FromResult(threadData.Keys.ToArray());
        }

        return Task.FromResult(Array.Empty<string>());
    }

    /// <inheritdoc/>
    public Task<bool> CheckHealthAsync()
    {
        return Task.FromResult(true);
    }
}

File: typescript/packages/agents-protocol-hosting/src/storage/InMemoryStorage.ts
import { IStorage } from './IStorage.js';

/**
 * In-memory storage implementation.
 * Data is lost when the process restarts.
 * Use only for development or testing.
 */
export class InMemoryStorage implements IStorage {
  private storage: Map<string, Map<string, unknown>> = new Map();

  /**
   * Creates a new in-memory storage instance.
   */
  constructor() {}

  async getAsync<T>(threadId: string, key: string): Promise<T | null> {
    const threadStorage = this.storage.get(threadId);
    if (!threadStorage) {
      return null;
    }

    const value = threadStorage.get(key);
    return value !== undefined ? (value as T) : null;
  }

  async setAsync<T>(threadId: string, key: string, value: T): Promise<void> {
    let threadStorage = this.storage.get(threadId);
    if (!threadStorage) {
      threadStorage = new Map<string, unknown>();
      this.storage.set(threadId, threadStorage);
    }

    threadStorage.set(key, value);
  }

  async deleteAsync(threadId: string, key: string): Promise<void> {
    const threadStorage = this.storage.get(threadId);
    if (threadStorage) {
      threadStorage.delete(key);
    }
  }

  async getKeysAsync(threadId: string): Promise<string[]> {
    const threadStorage = this.storage.get(threadId);
    if (!threadStorage) {
      return [];
    }

    return Array.from(threadStorage.keys());
  }

  async checkHealth(): Promise<boolean> {
    return true;
  }
}
